src/base/bittorrent/session.cpp

```diff
@@ -1277,36 +1294,65 @@ void Session::populateAdditionalTrackers()
-                Logger* const logger = Logger::instance();
-                if (maxRatioAction() == Remove) {
-                    logger->addMessage(tr("'%1' reached the maximum ratio you set. Removing...").arg(torrent->name()));
-                    deleteTorrent(torrent->hash());
+            if (torrent->ratioLimit() != TorrentHandle::NO_RATIO_LIMIT) {
+                const qreal ratio = torrent->realRatio();
+                qreal ratioLimit = torrent->ratioLimit();
+                if (ratioLimit == TorrentHandle::USE_GLOBAL_RATIO)
+                    // If Global Max Ratio is really set...
+                    ratioLimit = globalMaxRatio();
+
+                if (ratioLimit >= 0) {
+                    qDebug("Ratio: %f (limit: %f)", ratio, ratioLimit);
+                    Q_ASSERT(ratioLimit >= 0.f);
+
+                    if ((ratio <= TorrentHandle::MAX_RATIO) && (ratio >= ratioLimit)) {
+                        Logger* const logger = Logger::instance();
+                        if (m_maxRatioAction == Remove) {
+                            logger->addMessage(tr("'%1' reached the maximum ratio you set. Removing...").arg(torrent->name()));
+                            deleteTorrent(torrent->hash());
+                        }
+                        else {
+                            // Pause it
+                            if (!torrent->isPaused()) {
+                                logger->addMessage(tr("'%1' reached the maximum ratio you set. Pausing...").arg(torrent->name()));
+                                torrent->pause();
+                            }
+                        }
+                    }
                 }
-                else {
-                    // Pause it
-                    if (!torrent->isPaused()) {
-                        logger->addMessage(tr("'%1' reached the maximum ratio you set. Pausing...").arg(torrent->name()));
-                        torrent->pause();
+            }
+
+            if (torrent->seedingTimeLimit() != TorrentHandle::NO_SEEDING_TIME_LIMIT) {
+                const int seedingTime = torrent->seedingTime() / 60;
+                int seedingTimeLimit = torrent->seedingTimeLimit();
+                if (seedingTimeLimit == TorrentHandle::USE_GLOBAL_SEEDING_TIME)
+                     // If Global Seeding Time Limit is really set...
+                    seedingTimeLimit = globalMaxSeedingMinutes();
+
+                if (seedingTimeLimit >= 0) {
+                    qDebug("Seeding Time: %d (limit: %d)", seedingTime, seedingTimeLimit);
+                    Q_ASSERT(seedingTimeLimit >= 0);
+
+                    if ((seedingTime <= TorrentHandle::MAX_SEEDING_TIME) && (seedingTime >= seedingTimeLimit)) {
+                        Logger* const logger = Logger::instance();
+                        if (m_maxRatioAction == Remove) {
+                            logger->addMessage(tr("'%1' reached the maximum seeding time you set. Removing...").arg(torrent->name()));
```

Comment:
You should read the whole file not the diff.  This is a copy/paste from the same section that process the ratio limit that's already there.  I just tried to be consistent with the code that was already there.

This is YOUR code not mine:
```
                        if (m_maxRatioAction == Remove) {
                            logger->addMessage(tr("'%1' reached the maximum ratio you set. Removing...").arg(torrent->name()));
                            deleteTorrent(torrent->hash());
                        }
```
I just tried to be consistent in what was already there.  **DO NOT READ JUST DIFFS READ THE WHOLE FILE**